src/shared/wizards/wizardForm.ts

```diff
@@ -3,87 +3,159 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import { Prompter } from '../ui/prompter'
 import * as _ from 'lodash'
-import { StateWithCache, WizardState } from './wizard'
-import { ExpandWithObject } from '../utilities/tsUtils'
+import * as toposort from 'toposort'
+import { Prompter } from '../ui/prompter'
+import { StateWithCache } from './wizard'
+import { ExpandWithObject, ReduceTuple } from '../utilities/tsUtils'
+
+export type PrompterProvider<TState, TProp, TDep> = (
+    state: StateWithCache<BoundState<TState, TDep>, TProp>
+) => Prompter<TProp>
+export type BoundState<TState, Bindings> = FindPartialTuple<TState, ExtractPaths<Bindings>>
+
+type DefaultFunction<TState, TProp, TDep> = (state: BoundState<TState, TDep>) => TProp | undefined
+
+/**
+ * Changes all properties enumerated by `K` to be required while leaving everything else
+ * to be optional. For example, consider `K` as ['foo', 'bar'] for a `T` of:
+ * ```
+ * {
+ *    foo?: {
+ *       bar: string
+ *       qaz: string
+ *    }
+ *    baz: string
+ * }
+ * ```
+ * Would become:
+ * ```
+ * {
+ *    foo: {
+ *       bar: string
+ *       qaz?: string
+ *    }
+ *    baz?: string
+ * }
+ * ```
+ * `M` is used to describe when there is a matched key. Matched keys should return their type,
+ * while non-matches are considered to be unassigned, and thus are potentially undefined.
+ */
+type FindPartialTuple<T, K extends string[], M extends boolean = false> = K['length'] extends 0
+    ? M extends true
+        ? T
+        : RecursivePartial<T>
+    : {
+          [P in keyof T as P & K[0]]-?: FindPartialTuple<T[P], ReduceTuple<K>, true>
+      } &
+          {
+              [P in keyof T as Exclude<P, K[0]>]+?: FindPartialTuple<T[P], ReduceTuple<K>, false>
+          }
+
+type RecursivePartial<T> = {
+    [P in keyof Required<T>]+?: RecursivePartial<T[P]>
+}
 
-export type PrompterProvider<TState, TProp> = (state: StateWithCache<WizardState<TState>, TProp>) => Prompter<TProp>
+// This is basically just a 'type-helper' to retain information about where to find a property
+// relative to a 'root' element
+interface Pathable {
+    path: string[]
+}
 
-type DefaultFunction<TState, TProp> = (state: WizardState<TState>) => TProp | undefined
+// Extracts the 'paths' of a type relative to the root type
+type ExtractPaths<B> = B extends { path: infer K }[] ? (K extends string[] ? K : never) : never
```

Comment:
What is the expected output of this? At the end of all of this, it looks like you're just returning a series of arrays, is there an easier way to do this?